Translate ``cpu.frequency`` hardware requirement

skycastlelily opened the pull request:
Pull Request Checklist

* [x] implement the feature
* [ ] write the documentation
* [x] extend the test coverage
* [x] update the specification
* [ ] adjust plugin docstring
* [ ] modify the json schema
* [ ] mention the version
* [ ] include a release note

happz commented on spec/hardware/cpu.fmf:
```diff
@@ -41,6 +41,9 @@ description: |
           vendor: 1234|"> 1234"
 
+          # Integer or string, CPU speed.
+          speed: 2300|">= 2300"
```
What is "CPU speed"?

skycastlelily commented:
> What is "CPU speed"?

How quickly a cpu can process data, measured in HZ,ah,I should put this
information into spec?:)

On Thu, Oct 17, 2024 at 4:08 PM Miloš Prchlík ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In spec/hardware/cpu.fmf
> <https://github.com/teemtee/tmt/pull/3296#discussion_r1804306445>:
>
> > @@ -41,6 +41,9 @@ description: |
>            # Number or string, an ID of CPU vendor.
>            vendor: 1234|"> 1234"
>
> +          # Integer or string, CPU speed.
> +          speed: 2300|">= 2300"
>
> What is "CPU speed"?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/teemtee/tmt/pull/3296#pullrequestreview-2374472273>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKFR23D7EPAZZVPL6UR4GHDZ35V5VAVCNFSM6AAAAABQDDNFEGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDGNZUGQ3TEMRXGM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

martinhoyer commented:
> How quickly a cpu can process data, measured in HZ,ah,I should put this information into spec?:)

I believe in this context, we are using MHz.

skycastlelily commented:
Yeah, Mhz, Ghz, TBO, I don't use it for my team's testing, I'm just trying
to fill gaps.
My goal is: when beaker  users come , they find almost all they need are
already implemented☺️
Or, you think I should set a default unit for it?:)
Btw, for cpu.speed ,float is better,though int will fit,I guess, most of
users need, do you think I should add or extend
NumberConstraint(Constraint[int]) to have float support?

On Thu, Oct 17, 2024 at 5:42 PM Martin Hoyer ***@***.***>
wrote:

> How quickly a cpu can process data, measured in HZ,ah,I should put this
> information into spec?:)
>
> I believe in this context, we are using MHz.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/teemtee/tmt/pull/3296#issuecomment-2419055815>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKFR23FL32BS2ONAQPGZBOTZ36A67AVCNFSM6AAAAABQDDNFEGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDIMJZGA2TKOBRGU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

happz commented:
> > How quickly a cpu can process data, measured in HZ,ah,I should put this information into spec?:)
> 
> I believe in this context, we are using MHz. Also, I think it's best to do just Mhz to keep it simple? Or is there any use-case to do conversions et al, @happz ?

The good thing is, that we do not need to worry about conversions :) Thanks to Pint, `2.6 GHz` or `2400 MHz` should be easy to convert to the right unit when needed. Let the user input anything, make it clear the field expects hertz but GHz or kHz are perfectly fine, and convert it to whatever Beaker/AWS/Azure/GCP/whatever desires when constructing a recipe for that infrastructure (`value.to('GHz').magnitude`, for example).

It's like the `memory` key - user is not limited to one unit, as long as they stick to units describing a size of memory. And we say "Bytes are assumed when no unit is specified.". Let's do the same but with a unit one might use for CPU frequency, GHz, MHz, whatever we pick, as long as we let user know and use it accordingly in the code.

> Oh and it might be better to use "frequency", instead of "speed".

+1 for `frequency`, it's also clearer to me what it means. "speed" is pretty ambiguous, although it's widely used.

skycastlelily commented:
Updated^^

On Thu, Oct 17, 2024 at 6:15 PM Miloš Prchlík ***@***.***>
wrote:

> Yeah, Mhz, Ghz, TBO, I don't use it for my team's testing, I'm just trying
> to fill gaps. My goal is: when beaker users come , they find almost all
> they need are already implemented☺️ Or, you think I should set a default
> unit for it?:)
>
> Yep, see how memory is done.
>
> Btw, for cpu.speed ,float is better,though int will fit,I guess, most of
> users need, do you think I should add or extend
> NumberConstraint(Constraint[int]) to have float support?
>
> Hm, probably a new class should be added, and maybe we should rename
> NumberConstraint to something like IntegerConstraint, and use
> NumberConstraint for this new class which would run with float values. We
> still need the int one because cpu.processors or whatever.vendor ID will
> never be a floating-point number.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/teemtee/tmt/pull/3296#issuecomment-2419125561>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKFR23EZ6L7IGWVBW3VECDLZ36E2RAVCNFSM6AAAAABQDDNFEGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDIMJZGEZDKNJWGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

happz commented on docs/releases.rst:
```diff
@@ -18,7 +20,6 @@ option or with the environment variable ``TMT_FEELING_SAFE`` set
 to ``True``. See the :ref:`/stories/features/feeling-safe` section
 for more details and motivation behind this change.
 
-
```
Unrelated change.

happz commented on docs/releases.rst:
```diff
@@ -4,10 +4,12 @@
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+The :ref:`/plugins/provision/beaker` provision plugin gains support
+for :ref:`cpu.stepping</spec/hardware/cpu>` hardware requirements.
```
Singular requirement, "... hardware requirement.".

happz commented on tmt/hardware.py:
```diff
@@ -731,12 +731,50 @@ def _cast_number(raw_value: Any) -> int:
+            raw_value: str,
+            original_constraint: Optional['Constraint[Any]'] = None,
+            allowed_operators: Optional[list[Operator]] = None,
+            default_unit: Optional[Any] = 'MHz'
```
Most likely `None` would be better, `MHz` would make sense for just some of the requirements. Note that the original `NumberConstraint` did not have a default unit at all, it's just a dimensionless number most of the time. This one wouldn't be that different, and only *some* would need to specify this field.

happz commented on tmt/hardware.py:
```diff
@@ -904,11 +942,28 @@ def wrapper(spec: Spec, index: int) -> BaseConstraint:
+        ]
+
+
+def _parse_float_constraints(
```
Should be `_parse_number_constraints()`, to match the `NumberConstraint` class name.

happz commented on tmt/hardware.py:
```diff
@@ -1097,7 +1152,15 @@ def _parse_cpu(spec: Spec) -> BaseConstraint:
+        spec,
+        'cpu',
+        (
+            'frequency',
```
Here we would definitely specify the default unit, `MHz`, `_parse_number_constraints()` would then propagate it to `NumberConstraint` class. Note that this is similar to `memory`, but `memory` uses `SizeConstraint` where we do have a default unit already predefined, always `bytes`. For more generic "number" constraint, we can't predefined one, so we need to specify it when parsing a constraint.

happz commented on tmt/hardware.py:
```diff
@@ -731,12 +731,50 @@ def _cast_number(raw_value: Any) -> int:
+                raw_value = raw_value.strip()
+
+                if raw_value.startswith('0x'):
+                    return float(raw_value, base=16)
```
Relict from `int`, right? There's no `float` written in hexadecimal form.

happz commented on spec/hardware/cpu.fmf:
```diff
@@ -108,8 +114,13 @@ example:
+  - |
+    # Request a CPU with specified stepping.
+    cpu:
+        stepping: 9
```
Unrelated changes, `stepping`?

happz replied:
I see, stepping commits are included in this PR. You can edit this PR to change its base branch to the one providing stepping changes, that should remove stepping changes from this PR's diff.

happz commented on spec/hardware/cpu.fmf:
```diff
@@ -108,8 +114,13 @@ example:
     note: "``cpu.vendor``, ``cpu.vendor-name`` and ``cpu.hyper-threading`` not implemented yet"
   - implemented-by: /tmt/steps/provision/mrack.py
-    note: "``cpu.flag``, ``cpu.processors``, ``cpu.model``, ``cpu.cores``, ``cpu.model-name``, ``cpu.hyper-threading`` and ``cpu.vendor-name`` only"
+    note: "``cpu.flag``, ``cpu.processors``, ``cpu.model``, ``cpu.cores``, ``cpu.model-name``, ``cpu.hyper-threading``, ``cpu.stepping`` and ``cpu.vendor-name`` only"
```
Doesn't mention `frequency`.

happz commented on spec/hardware/cpu.fmf:
```diff
@@ -41,6 +41,9 @@ description: |
           # Number or string, an ID of CPU vendor.
           vendor: 1234|"> 1234"
 
+          # Float or string, CPU frequency.
```
Doesn't mention the default unit, see how `memory` is documented.

skycastlelily commented:
That's because this is cpu-speed branch, cpu.frequency for mrack is
implemented on cpu-speed-mrack branch, which is based on this branch:)

On Fri, Oct 18, 2024 at 5:06 PM Miloš Prchlík ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In spec/hardware/cpu.fmf
> <https://github.com/teemtee/tmt/pull/3296#discussion_r1806156666>:
>
> >  link:
>    - implemented-by: /tmt/steps/provision/artemis.py
>      note: "``cpu.vendor``, ``cpu.vendor-name`` and ``cpu.hyper-threading`` not implemented yet"
>    - implemented-by: /tmt/steps/provision/mrack.py
> -    note: "``cpu.flag``, ``cpu.processors``, ``cpu.model``, ``cpu.cores``, ``cpu.model-name``, ``cpu.hyper-threading`` and ``cpu.vendor-name`` only"
> +    note: "``cpu.flag``, ``cpu.processors``, ``cpu.model``, ``cpu.cores``, ``cpu.model-name``, ``cpu.hyper-threading``, ``cpu.stepping`` and ``cpu.vendor-name`` only"
>
> Doesn't mention frequency.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/teemtee/tmt/pull/3296#pullrequestreview-2377491819>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKFR23DSB55HXZ4P54HGUPDZ4DFRNAVCNFSM6AAAAABQDDNFEGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDGNZXGQ4TCOBRHE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

skycastlelily commented:
>  You can edit this PR to change its base branch to the one providing
stepping changes

Updated and changed^^

On Fri, Oct 18, 2024 at 11:12 PM Lili Nie ***@***.***> wrote:

> That's because this is cpu-speed branch, cpu.frequency for mrack is
> implemented on cpu-speed-mrack branch, which is based on this branch:)
>
> On Fri, Oct 18, 2024 at 5:06 PM Miloš Prchlík ***@***.***>
> wrote:
>
>> ***@***.**** commented on this pull request.
>> ------------------------------
>>
>> In spec/hardware/cpu.fmf
>> <https://github.com/teemtee/tmt/pull/3296#discussion_r1806156666>:
>>
>> >  link:
>>    - implemented-by: /tmt/steps/provision/artemis.py
>>      note: "``cpu.vendor``, ``cpu.vendor-name`` and ``cpu.hyper-threading`` not implemented yet"
>>    - implemented-by: /tmt/steps/provision/mrack.py
>> -    note: "``cpu.flag``, ``cpu.processors``, ``cpu.model``, ``cpu.cores``, ``cpu.model-name``, ``cpu.hyper-threading`` and ``cpu.vendor-name`` only"
>> +    note: "``cpu.flag``, ``cpu.processors``, ``cpu.model``, ``cpu.cores``, ``cpu.model-name``, ``cpu.hyper-threading``, ``cpu.stepping`` and ``cpu.vendor-name`` only"
>>
>> Doesn't mention frequency.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/teemtee/tmt/pull/3296#pullrequestreview-2377491819>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AKFR23DSB55HXZ4P54HGUPDZ4DFRNAVCNFSM6AAAAABQDDNFEGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDGNZXGQ4TCOBRHE>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>

happz commented:
Blocked on https://github.com/teemtee/tmt/pull/3295

happz commented on tmt/hardware.py:
```diff
@@ -731,12 +731,46 @@ def _cast_number(raw_value: Any) -> int:
 
 
+class NumberConstraint(Constraint[float]):
+    """ A constraint representing a dimension-less float number """
```
`IntegerConstraint` remains dimension-less, but `NumberConstraint` isn't, we already use it for a constraint with a default unit :)

skycastlelily commented:
Ah,yes,and updated^^

On Thu, Oct 24, 2024 at 4:09 PM Miloš Prchlík ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In tmt/hardware.py
> <https://github.com/teemtee/tmt/pull/3296#discussion_r1814489115>:
>
> >              original_constraint=original_constraint,
>              allowed_operators=allowed_operators
>              )
>
>
> +class NumberConstraint(Constraint[float]):
> +    """ A constraint representing a dimension-less float number """
>
> IntegerConstraint remains dimension-less, but NumberConstraint isn't, we
> already use it for a constraint with a default unit :)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/teemtee/tmt/pull/3296#pullrequestreview-2391657432>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKFR23EPGMPDG7MGQVCIYFDZ5CTKFAVCNFSM6AAAAABQDDNFEGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDGOJRGY2TONBTGI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

happz commented:
Unrelated Fedora 41 failures, merging.
